dm/worker/task_checker.go

```diff
@@ -0,0 +1,315 @@
+	if stStatus == nil || stStatus.Stage != pb.Stage_Paused || stStatus.Result == nil || stStatus.Result.IsCanceled {
+		return ResumeIgnore
+	}
+
+	// TODO: use different strategies based on the error detail
+	for _, processErr := range stStatus.Result.Errors {
+		if !isResumableError(processErr) {
+			return ResumeNoSense
+		}
+	}
+
+	// auto resume interval does not exceed backoff duration, skip this paused task
+	if time.Since(tsc.bc.latestResumeTime[stStatus.Name]) < duration {
+		return ResumeSkip
+	}
+
+	return ResumeDispatch
+}
+
+func (tsc *realTaskStatusChecker) check() {
+	allSubTaskStatus := tsc.w.getAllSubTaskStatus()
+
+	defer func() {
+		// cleanup outdated tasks
+		for taskName := range tsc.bc.backoffs {
+			_, ok := allSubTaskStatus[taskName]
+			if !ok {
+				tsc.l.Debug("remove task from checker", zap.String("task", taskName))
+				delete(tsc.bc.backoffs, taskName)
+				delete(tsc.bc.latestPausedTime, taskName)
+				delete(tsc.bc.latestBlockTime, taskName)
+				delete(tsc.bc.latestResumeTime, taskName)
+			}
+		}
+	}()
+
+	for taskName, stStatus := range allSubTaskStatus {
+		bf, ok := tsc.bc.backoffs[taskName]
+		if !ok {
+			bf, _ = backoff.NewBackoff(tsc.cfg.BackoffFactor, tsc.cfg.BackoffJitter, tsc.cfg.BackoffMin, tsc.cfg.BackoffMax)
+			tsc.bc.backoffs[taskName] = bf
+			tsc.bc.latestPausedTime[taskName] = time.Now()
+			tsc.bc.latestResumeTime[taskName] = time.Now()
+		}
+		duration := bf.Current()
+		strategy := tsc.getResumeStrategy(stStatus, duration)
+		switch strategy {
+		case ResumeIgnore:
+			if time.Since(tsc.bc.latestPausedTime[taskName]) > tsc.cfg.BackoffRollback {
+				bf.Rollback()
+				// after each rollback, reset this timer
+				tsc.bc.latestPausedTime[taskName] = time.Now()
+			}
+		case ResumeNoSense:
+			// this strategy doesn't forward or rollback backoff
+			tsc.bc.latestPausedTime[taskName] = time.Now()
+			blockTime, ok := tsc.bc.latestBlockTime[taskName]
+			if ok {
+				tsc.l.Warn("task can't auto resume", zap.String("task", taskName), zap.Duration("paused duration", time.Since(blockTime)))
+			} else {
+				tsc.bc.latestBlockTime[taskName] = time.Now()
+				tsc.l.Warn("task can't auto resume", zap.String("task", taskName))
+			}
+		case ResumeSkip:
+			tsc.l.Warn("backoff skip auto resume task", zap.String("task", taskName), zap.Time("latestResumeTime", tsc.bc.latestResumeTime[taskName]), zap.Duration("duration", duration))
+			tsc.bc.latestPausedTime[taskName] = time.Now()
+		case ResumeDispatch:
+			tsc.bc.latestPausedTime[taskName] = time.Now()
+			opLogID, err := tsc.w.operateSubTask(&pb.TaskMeta{
+				Name: taskName,
+				Op:   pb.TaskOp_AutoResume,
+			})
+			if err != nil {
+				tsc.l.Error("dispatch auto resume task failed", zap.String("task", taskName), zap.Error(err))
+			} else {
+				tsc.l.Info("dispatch auto resume task", zap.String("task", taskName), zap.Int64("opLogID", opLogID))
+				tsc.bc.latestResumeTime[taskName] = time.Now()
+				bf.Forward()
```

Comment:
if called `Forward` many times because of auto resume not take effects, then `cwnd` can become very large.
later, we need to call `Rollback` multi times to make `minf * math.Pow(b.Factor, float64(cwnd)) < b.Max`, right?